### PR TITLE
UCT/IB/MLX5: Refactor KSM functions and extend logging

### PR DESCRIPTION
**Pull request by yosefe**
## Why
Improve KSM debug logging

### CODE REVIEW

**Comment by brminich** [comment]
seems it affects the code which is significantly modified in #9424

**Comment by yosefe** [comment]
@ivankochin @brminich can you pls take a look?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 1]
```diff
@@ -186,23 +199,35 @@ uct_ib_mlx5_devx_reg_ksm_data_addr(uct_ib_mlx5_md_t *md, struct ibv_mr *mr,
...
+    void *klm;
+    char *in;
+    int i;
+
+    if (reason != NULL) {
+        ucs_trace("%s: initialize ksm_data_addr list for %s address 0x%lx "
+                  "length %zu iova 0x%" PRIx64,
+                  uct_ib_device_name(&md->super.dev), reason, address, length,
+                  iova);
+    }
```
Comment (brminich) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:213]: minor: can be inline func and reused also in `uct_ib_mlx5_devx_reg_ksm_data_mt` (together with check `reason != NULL`)

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 2]
```diff
@@ -140,9 +141,16 @@ uct_ib_mlx5_devx_reg_ksm_data_mt(uct_ib_mlx5_md_t *md, int atomic,
     void *mr_address       = address;
     size_t list_size       = ksm_data->mr_num;
     ucs_status_t status;
+    struct ibv_mr **mr;
     char *in;
     void *klm;
-    int i;
+
+    if (reason != NULL) {
```
Comment (ivankochin) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:148]: I think it worth printing this information even if there is no reason defined. WDYT?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 3]
```diff
@@ -154,13 +162,18 @@ uct_ib_mlx5_devx_reg_ksm_data_mt(uct_ib_mlx5_md_t *md, int atomic,
         goto out;
     }
 
+    ucs_log_indent(+1);
     klm = UCT_IB_MLX5DV_ADDR_OF(create_mkey_in, in, klm_pas_mtt);
-    for (i = 0; i < ksm_data->mr_num; i++) {
+    ucs_carray_for_each(mr, ksm_data->mrs, ksm_data->mr_num) {
+        ucs_trace("klm[%ld] va %p mr [addr %p len %zu lkey 0x%x]",
```
Comment (ivankochin) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:168]: Seems like this trace is the same with one inside `uct_ib_mlx5_devx_reg_ksm_data_addr`. Maybe it worth to remove code duplication by asserts to separate functions (e.g. `uct_ib_mlx5_devx_klm_entry_log`)?

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 4]
```diff
@@ -216,12 +239,11 @@ uct_ib_mlx5_devx_reg_ksm_data_addr(uct_ib_mlx5_md_t *md, struct ibv_mr *mr,
  * that were registered in single-threaded mode.
  */
 static ucs_status_t uct_ib_mlx5_devx_reg_ksm_data_contig(
```
Comment (ivankochin) [inline src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:241]: I have collected all the ksm registration related functions signatures and detected some inconsistency in them. E.g. all the functions took `atomic` argument after `iova` or `iova_offset` argument but `uct_ib_mlx5_devx_reg_ksm` took it right after `address`. Here is the current API aligned by arguments:

```c
uct_ib_mlx5_devx_reg_ksm            (uct_ib_mlx5_md_t *md, int atomic, uint64_t address, size_t length, int list_size, size_t entity_size, char *in,         uint32_t mkey_index, const char *reason,                                                                                 struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_ksm_log            (uct_ib_mlx5_md_t *md,                void *address, size_t length,            uint64_t iova,        int atomic, int mt, uint32_t mkey_index, const char *reason, ucs_status_t status)
uct_ib_mlx5_devx_reg_ksm_data_mt    (uct_ib_mlx5_md_t *md,                void *address,                           uint64_t iova,        int atomic,         uint32_t mkey_index, const char *reason, uct_ib_mlx5_devx_ksm_data_t *ksm_data,                                          struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_reg_ksm_data_addr  (uct_ib_mlx5_md_t *md,                void *address,                           uint64_t iova,        int atomic,         uint32_t mkey_index, const char *reason, struct mlx5dv_devx_obj **mr_p, struct ibv_mr *mr, size_t length, int list_size, struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_reg_ksm_data_contig(uct_ib_mlx5_md_t *md,                void *address,                           uint32_t iova_offset, int atomic,         uint32_t mkey_index, const char *reason, uct_ib_mlx5_devx_mr_t *mr,                                                      struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_reg_ksm_data       (uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh, uct_ib_mr_type_t mr_type, uint32_t iova_offset, int atomic,         uint32_t mkey_index, const char *reason,                                                                                 struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
```

I suggest to align all the interfaces by one pattern of arguments, e.g.:
```c
uct_ib_mlx5_devx_ksm_log            (uct_ib_mlx5_md_t *md,                void *address, size_t length,            uint64_t iova,        int atomic, uint32_t mkey_index, const char *reason, int mt, ucs_status_t status)
uct_ib_mlx5_devx_reg_ksm            (uct_ib_mlx5_md_t *md,             uint64_t address, size_t length,                                  int atomic, uint32_t mkey_index, const char *reason, int list_size, size_t entity_size, char *in,                     struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_reg_ksm_data_mt    (uct_ib_mlx5_md_t *md,                void *address,                           uint64_t iova,        int atomic, uint32_t mkey_index, const char *reason, uct_ib_mlx5_devx_ksm_data_t *ksm_data,                           struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_reg_ksm_data_addr  (uct_ib_mlx5_md_t *md,                void *address, size_t length,            uint64_t iova,        int atomic, uint32_t mkey_index, const char *reason, struct mlx5dv_devx_obj **mr_p, struct ibv_mr *mr, int list_size, struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_reg_ksm_data_contig(uct_ib_mlx5_md_t *md,                void *address,                           uint32_t iova_offset, int atomic, uint32_t mkey_index, const char *reason, uct_ib_mlx5_devx_mr_t *mr,                                       struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
uct_ib_mlx5_devx_reg_ksm_data       (uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh, uct_ib_mr_type_t mr_type, uint32_t iova_offset, int atomic, uint32_t mkey_index, const char *reason,                                                                  struct mlx5dv_devx_obj **mr_p, uint32_t *mkey)
```

In the proposed change the arguments in all the functions are placed in the following order:
1. MD
2. Info required to get address + length
3. iova or iova_offset
4. atomic
5. mkey_index
6. reason
7. MRs related info
8. unique arguments
9. devx output structure pointer
10. mkey output pointer

**Comment by ivankochin** [comment]
BTW it seems like the title of this PR doesn't fully represent the content anymore.

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 5]
```diff
@@ -112,102 +110,127 @@ uct_ib_mlx5_devx_reg_ksm(uct_ib_mlx5_md_t *md, int atomic, uint64_t address,
...
-    ucs_debug("KSM %s-thread memory registration status \"%s\" "
-              "range %p..%p iova 0x%" PRIx64 "%s on %s",
-              mt ? "multi" : "single", ucs_status_string(status), address,
+    ucs_debug("%s: KSM%s %s memory registration status \"%s\" "
+              "range %p..%p iova 0x%" PRIx64 "%s mkey_index 0%x",
+              uct_ib_device_name(&md->super.dev), mt ? "-mt" : "", reason,
+              ucs_status_string(status), address,
               UCS_PTR_BYTE_OFFSET(address, length), iova,
-              atomic ? " atomic" : "", uct_ib_device_name(&md->super.dev));
+              atomic ? " atomic" : "", mkey_index);
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:123]: why is it better? I'd prefer to reduce size of constant strings

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c [hunk 6]
```diff
@@ -112,102 +110,127 @@ uct_ib_mlx5_devx_reg_ksm(uct_ib_mlx5_md_t *md, int atomic, uint64_t address,
...
+                                     size_t length, uint64_t iova, int atomic,
+                                     int mt, uint32_t mkey_index,
+                                     const char *reason, ucs_status_t status)
 {
-    ucs_debug("KSM %s-thread memory registration status \"%s\" "
-              "range %p..%p iova 0x%" PRIx64 "%s on %s",
-              mt ? "multi" : "single", ucs_status_string(status), address,
+    ucs_debug("%s: KSM%s %s memory registration status \"%s\" "
+              "range %p..%p iova 0x%" PRIx64 "%s mkey_index 0%x",
+              uct_ib_device_name(&md->super.dev), mt ? "-mt" : "", reason,
```
[the remark this answers is not in the document]
Comment (yosefe) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:120]: reason should not be NULL anymore after this PR

File: src/uct/ib/mlx5/dv/ib_mlx5dv_md.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (ivankochin) [reply src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:124]: Sorry, my bad, this change doesn't have a sense

**Review by ivankochin — APPROVED** [review]

**Pull request merged**